Monitor pods by labels instead of names

dimberman opened the pull request:
Make sure you have checked _all_ steps below.

### Jira

- [x] My PR addresses the following [Airflow Jira](https://issues.apache.org/jira/browse/AIRFLOW/) issues and references them in the PR title. For example, "\[AIRFLOW-XXX\] My Airflow PR"
  - https://issues.apache.org/jira/browse/AIRFLOW-5589
  - In case you are fixing a typo in the documentation you can prepend your commit with \[AIRFLOW-XXX\], code changes always need a Jira issue.
  - In case you are proposing a fundamental code change, you need to create an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)).
  - In case you are adding a dependency, check if the license complies with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

To prevent situations where the scheduler starts a
second k8sPodOperator pod after a restart, we now check
for existing pods using kubernetes labels
### Tests

- [x] My PR adds the following unit tests __OR__ does not need testing for this extremely good reason:

### Commits

- [x] My commits all reference Jira issues in their subject lines, and I have squashed multiple commits if they address the same issue. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation

- [x] In case of new functionality, my PR adds documentation that describes how to use it.
  - All the public functions and the classes in the PR contain docstrings that explain what it does
  - If you implement backwards incompatible changes, please leave a note in the [Updating.md](https://github.com/apache/airflow/blob/master/UPDATING.md) so we can assign it to a appropriate release

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -23,6 +25,28 @@
 from airflow.utils.state import State
 
 
+def create_labels_for_pod(context):
```
Make this a `@staticmethod` of KubernetesPodOperator?

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -23,6 +25,28 @@
+    }
+    # In the case of sub dags this is just useful
+    if context['dag'].parent_dag:
+        labels['parent_dag_id'] = context['dag'].parent_dag.dag_id
```
Do we need this or is it just "nice" as it makes it easier to see what's going on?

dimberman replied:
I'm thinking it's just a "nice" thing to have. Unless there's a rare race condition of people subDagging the same DAG at exactly the same time (which I highly doubt)

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -23,6 +25,28 @@
+        labels['parent_dag_id'] = context['dag'].parent_dag.dag_id
+    # Replace unsupported characters with dashes & trim at 63 chars (max allowed)
+    for label_id, label in labels.items():
+        label = re.sub(r'[^-A-Za-z0-9_.]+', '-', label)[:63]
```
We should refactor `_make_safe_label_value` from KubeExcutor and move it to `make_safe_label_value` on PodGenerator then call it form both places.

dimberman replied:
Agreed

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -112,55 +113,54 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
-                                                extract_xcom=self.do_xcom_push)
+            if len(pod_list.items) == 1:
+                pod = pod_list.items[0]
+                final_state, result = self.monitor_launched_pod(launcher, pod)
```
Add an info log line here about found/monitoring existing pod

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -112,55 +113,54 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
+        labels = {
+            'dag_id': context['dag'].dag_id,
+            'task_id': context['task'].task_id,
+            'exec_date': context['ts']
```
We should include the try_number in here too - (i.e if try 1 is still running somehow but we want to launch try 2 we don't want to monitor try 1 again.)

dimberman replied:
Is that something that could happen? Would the execution date be the same?

ashb replied:
Yeah, if the task fails and it retires :) (It's unlikely to still be running, but some error/odd behavoiur could make it happen, but the "key" that uniqueue identifies a Task Instance is (dag_id,task_id, execution_date,try_number)

danccooper replied:
Hi guys,

I think we should consider the desired behaviour with this.  If we're including the try_number then it opens a situation where multiple tries could be running alongside each other.  I don't think that is ever desired & would certainly cause problems with my jobs (back to the original cause of this one 😄).

If we want to avoid monitoring the previous try, then should we check for it & kill it if exists before starting the next?

alistardust replied:
@danccooper @ashb Question about this, I think retry_number would actually be good here... Especially if you are debugging and [kubernetes][delete_worker_pods] is "False". Seems like it would delete the original pod under @danccooper's suggestion and you'd lose that for debugging purposes.

dakov replied:
On our k8s cluster, we keep pods in the namespace for some time no matter the state they finished in (success, OOMKilled, evicted,....) -- I do not maintain the cluster and the decision is not up to me. Not having try_number as an identifying label of a pod makes airflow fail to launch another attempt.

danccooper replied:
Hey @dakov I was going to suggest you just set the reattach_on_restart flag to False, however from checking the code I'm not sure this will work as expected which is probably the behaviour you're reporting? After discussing wth @kaxil he has rasied https://github.com/apache/airflow/issues/10021 to cover this, suggest we continue discussion on there.

mik-laj commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -171,6 +171,69 @@ def execute(self, context):
+    def create_new_pod_for_operator(self, labels, launcher):
+        """
+        Create a new pod and monitor it to completion
+        :param labels:
```
```suggestion

                :param labels:
```
Otherwise, docs doesn't looks good.
https://github.com/apache/airflow/pull/6344#discussion_r336709991

mik-laj commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -112,55 +113,54 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
+    def create_labels_for_pod(context):
+        """
+        Generate labels for the pod s.t. we can track it in case of Operator crash
+        :param context:
```
```suggestion

        :param context:
```

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -112,55 +113,55 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
+        # In the case of sub dags this is just useful
+        if context['dag'].parent_dag:
+            labels['parent_dag_id'] = context['dag'].parent_dag.dag_id
+        # Replace unsupported characters with dashes & trim at 63 chars (max allowed)
```
This comment probably doesn't reflect what make_safe_label_value does anymore so
```suggestion
        # Ensure that label is valid for Kube, and if not trucate/remove invalid chars and replace with short hash.
```

ashb commented on tests/integration/kubernetes/test_kubernetes_pod_operator.py:
```diff
@@ -61,7 +72,10 @@ def setUp(self):
                 'annotations': {},
-                'labels': {'foo': 'bar'}
+                'labels': {'foo': 'bar',
+                           'exec_date': '2019-01-01T0900000000-4bf747daa',
```
I wonder if it's worth using a non-standard date format here, rather than always hashing this label? (Ultimately it doesn't matter much to us as Airflow I guess.)

ashb commented on tests/integration/kubernetes/test_kubernetes_pod_operator.py:
```diff
@@ -51,6 +53,49 @@
+    def compare_two_pods(self, actual_pod):
+        """
+        This is a disgusting hack that is unfortunately necessary because kubernetes is not consistent
+        in it's return objects between "create_pod" and "get_pod"
```
I wouldn't say that it's not consistent, it's more that once the container is running then more of the fields are filled out.

ashb commented on tests/integration/kubernetes/test_kubernetes_pod_operator.py:
```diff
@@ -51,6 +53,49 @@
+        self._compare_sub_dicts(filtered_spec_metadata, expected_sub_pod=actual_pod['spec'])
+
+    def _compare_pod_containers(self, actual_containers, expected_containers):
+        container_map = {container['name']: container for container in expected_containers}
```
The length (already tested) and order of the containers matters to, so we should probably just do the comapre like:

```
    for (expected, actual) in zip(expected_containers, actual_containers):
        self._compare_sub_dicts(expected, actual)
```

I think?

dimberman replied:
Does the order matter? Yaml is declarative so I think the general assumption is that all containers are ready "when they're ready." The only place I'd imagine order mattering is between init-container -> main containers

ashb replied:
It's an array of dicts, so yes the order matters. The first container is also what we consider the "main" container

mik-laj replied:
Why do we recognize by order instead of by name?  In one place I saw that it is recognized by its name. I proposed a small change of this approach in my PR. https://github.com/apache/airflow/pull/6375/files#diff-0e7e9d15b866a8c22f75dd8aa447a54bR153

danccooper commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -112,55 +113,60 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
+            'dag_id': context['dag'].dag_id,
+            'task_id': context['task'].task_id,
+            'exec_date': context['ts'],
+            'try_number': context['ti'].try_number,
```
I think we need to consider the desired behaviour here.

What could happen now is that we create duplicate pods in the case of a previous try still running e.g. in the case of task timeout & no kill or failed kill.  In that case we introduce a new scenario of duplicate pods which is what we're trying to solve.  Certainly for my jobs this would cause issues as we're not using try_number to create output locations for example.

One option is to check for the previous try pod first and kill it if it exists before running the next one.

ashb replied:
The counter argument to this: if we _don't_ include try_number here then it could "reattach" to a previous running pod when Airflow thinks it has retired. That seems like an odd behaviour.

This is all edge cases anyway, it's unlikely that we would be in this situation.

What we could do is search by dag_id/task_id/exec_date (i.e. don't search by try_number) and then if we find a running pod with something other than the expected try_number we kill it? A little bit worried we might kill the wrong pod somehow there, but it _should_ be safe?

danccooper replied:
I agree with all that you've said, I think that is a good solution and should be safe.

ashb replied:
I'm now second guessing myself on this.

If we do add this check on try_number in I can't think of a case when we would re-attach to a running pod. (because if this operator fails/dies/zombies for it to run again it would be the next try_number.)

danccooper replied:
I should've realised that.  We do indeed monitor a previous try number on purpose in my original changes.

With the try number changes the main duplicate pod issue will be fixed still as we force kill it before starting a new one.  But I took the view that, although perhaps confusing in terms of try number, it is far more efficient to continue monitoring the previous try in the case of worker failure.  AFAIK worker failure/restart is the only situation in which pods remain running and are not killed.  If you fail a task via the UI it triggers a pod kill for example.

Unless you have another solution that can check & re-use the previous try number then I think we're better off removing the try_number changes?

danccooper replied:
Hey @ashb @dimberman how do feel about removing the try_number changes and allowing the re-attach to a previous pod?  This is the logic I've been running for a while now.  The situation I can think of where you'd not want this is if the pod hangs for some reason (sub process failure or similar) and then a kill on timeout fails.  Quite unlikely... but to consider.

alistardust replied:
@danccooper I've definitely run into that scenario

danccooper commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -112,55 +113,61 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
+            launcher = pod_launcher.PodLauncher(kube_client=client, extract_xcom=self.do_xcom_push)
+
+            if len(pod_list.items) == 1 and self._try_numbers_do_not_match(context, pod_list[0]):
+                self.log.info("found a running pod with labels %s."
```
@dimberman I think the logic looks sounds, you just need to shift this log message into the 'elif 1' & I guess we want to info here that we've found a previous pod & will kill it

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -112,55 +113,61 @@ class KubernetesPodOperator(BaseOperator):  # pylint: disable=too-many-instance-
+            'dag_id': context['dag'].dag_id,
+            'task_id': context['task'].task_id,
+            'exec_date': context['ts'],
+            'try_number': context['ti'].try_number,
```
```suggestion
            'try_number': str(context['ti'].try_number),
```

might be needed?

ashb replied:
Oh I see you've done it else where.

ashb commented on airflow/contrib/operators/kubernetes_pod_operator.py:
```diff
@@ -171,6 +178,76 @@ def execute(self, context):
             raise AirflowException('Pod Launching failed: {error}'.format(error=ex))
 
+    def _try_numbers_do_not_match(self, context, pod):
+        return pod['metadata']['labels']['task_instance']['try_number'] != context['ti'].try_number
```
This is never going to match as the types are different - string vs int

ashb commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -235,84 +259,46 @@ def execute(self, context):
+                self.log.info("found a running pod with labels %s."
+                              "Will monitor instead of starting new one", labels)
+                launcher.delete_pod(pod_list[0])
+                final_state, _, result = self.create_new_pod_for_operator(labels, launcher)
```
This log message doesn't line up with what the next two lines are doing.

dimberman replied:
moved it

ashb commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -324,3 +310,97 @@ def _set_name(self, name):
+            if self.is_delete_operator_pod:
+                launcher.delete_pod(pod)
+
+        if final_state != State.SUCCESS:
```
The event fetching should live in monitor_launched_pod, otherwise if we "attach"/monitor an existing pod we'll never check events for it.

ashb replied:
Ideally also can we change how we call launcher so that this fn (`create_new_pod_for_operator`) can call `monitor_launched_pod` -- so that both paths (creats vs attach) have the same behavoiur?

(Right now it appears the l.run_pod is blocking?)

ashb replied:
This probably means splitting `run_pod` inside launcher to be (pseudo code)

```python
pod = create_pod(...)
return monitor(pod)
```

kaxil commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -66,6 +70,25 @@ class PodDefaults:
+    end with an alphanumeric character ([a-z0-9A-Z]) with dashes (-), underscores (_),
+    dots (.), and alphanumerics between.
+
+    If the label value is then greater than 63 chars once made safe, or differs in any
```
```suggestion
    If the label value is greater than 63 chars once made safe, or differs in any
```

kaxil commented on airflow/kubernetes/pod_generator.py:
```diff
@@ -66,6 +70,25 @@ class PodDefaults:
+
+    If the label value is then greater than 63 chars once made safe, or differs in any
+    way from the original value sent to this function, then we need to truncate to
+    53chars, and append it with a unique hash.
```
```suggestion
    53 chars, and append it with a unique hash.
```

kaxil commented on airflow/kubernetes/pod_launcher.py:
```diff
@@ -116,9 +114,15 @@ def run_pod(
+        :param pod:
+        :@type V1Pod:
+        :param get_logs:
+        :return:
```
Can we fill the info here !

ashb commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -235,84 +259,45 @@ def execute(self, context):
-                    launcher.delete_pod(pod)
+            launcher = pod_launcher.PodLauncher(kube_client=client, extract_xcom=self.do_xcom_push)
 
+            if len(pod_list.items) == 1 and self._try_numbers_do_not_match(context, pod_list.items[0]):
```
So to meet danccooper's original ask: He want's a way of "resuming" for an old pod from an previous try.

For example he has a Kube pod that takes 30 hours to run, if Airflow "retries" the task, he want's it to notice and not kill the pod, but instead retry it.

I think that _might_ be unepxected, so a flag to KubePodOp of, for example`reattach_previous_pod_from_previous_try` (thought that is rather longer than I would like) is the way to go I think.

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -225,7 +225,31 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
+        Generate labels for the pod s.t. we can track it in case of Operator crash
+
+        :param context:
+        :return:
```
Here too

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -225,7 +225,31 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
+    @staticmethod
+    def create_labels_for_pod(context):
+        """
+        Generate labels for the pod s.t. we can track it in case of Operator crash
```
```suggestion
        Generate labels for the pod to track the pod in case of Operator crash
```

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -225,7 +225,31 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
+        labels = {
+            'dag_id': context['dag'].dag_id,
+            'task_id': context['task'].task_id,
+            'exec_date': context['ts'],
```
```suggestion
            'execution_date': context['ts'],
```

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -225,7 +225,31 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
+            'try_number': context['ti'].try_number,
+        }
+        # In the case of sub dags this is just useful
+        if context['dag'].parent_dag:
```
Would `dag.is_subdag` be a better check ?

```suggestion
        if context['dag'].is_subdag:
```

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -225,7 +225,31 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
+        if context['dag'].parent_dag:
+            labels['parent_dag_id'] = context['dag'].parent_dag.dag_id
+        # Ensure that label is valid for Kube,
+        # and if not trucate/remove invalid chars and replace with short hash.
```
```suggestion
        # and if not truncate/remove invalid chars and replace with short hash.
```

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -235,84 +259,45 @@ def execute(self, context):
 
+    @staticmethod
+    def _get_pod_identifying_label_string(labels):
+        filtered_labels = {label_id: label for label_id, label in labels.items() if label_id != 'try_number'}
```
Is it not possible to use multiple labels to get the pod instead of creating a single string?

If not, we need to be careful as the order of items can change how the string. We can `sort` the keys in `filtered_labels` to take care of that issue.

dimberman replied:
@kaxil why do we need to sort? this is for creating the label_selector for kubernetes which I don't believe is order specific.

kaxil replied:
For the following reason:

```
In [1]: def _get_pod_identifying_label_string(labels):
   ...:         filtered_labels = {label_id: label for label_id, label in labels.items() if label_id != 'try_number'}
   ...:         return ','.join([label_id + '=' + label for label_id, label in filtered_labels.items()])
   ...:

In [2]: labels = {
   ...:             'dag_id': 'test_dag_id',
   ...:             'task_id': 'test_task_id',
   ...:             'execution_date': 'execution_date',
   ...:             'try_number': 2,
   ...:         }

In [3]: _get_pod_identifying_label_string(labels)
Out[3]: 'dag_id=test_dag_id,task_id=test_task_id,execution_date=execution_date'

In [6]: labels_1 = {
   ...:             'task_id': 'test_task_id',
   ...:             'dag_id': 'test_dag_id',
   ...:             'execution_date': 'execution_date',
   ...:             'try_number': 2,
   ...:         }

In [7]: _get_pod_identifying_label_string(labels_1)
Out[7]: 'task_id=test_task_id,dag_id=test_dag_id,execution_date=execution_date'

```

ashb reviewed (dismissed):
Please add docs to the new `reattach_on_restart` parameter

astro-sql-decorator committed:
Monitor k8sPodOperator pods by labels

To prevent situations where the scheduler starts a
second k8sPodOperator pod after a restart, we now check
for existing pods using kubernetes labels

kaxil commented on airflow/kubernetes/pod_launcher.py:
```diff
@@ -116,9 +114,15 @@ def run_pod(
-    def _monitor_pod(self, pod: V1Pod, get_logs: bool) -> Tuple[State, Optional[str]]:
+        :param pod:
+        :@type V1Pod:
+        :param get_logs:
```
Needs description

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -225,7 +229,31 @@ def __init__(self,  # pylint: disable=too-many-arguments,too-many-locals
+        """
+        Generate labels for the pod to track the pod in case of Operator crash
+
+        :param context:
```
Needs a description or just remove it

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -324,3 +316,99 @@ def _set_name(self, name):
+
+        @param labels:
+        @param launcher:
+        @return:
```
same here

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -324,3 +316,99 @@ def _set_name(self, name):
         return re.sub(r'[^a-z0-9.-]+', '-', name.lower())
+
+    def create_new_pod_for_operator(self, labels, launcher) -> Tuple[State, k8s.V1Pod, Optional[str]]:
+        '''
```
```suggestion
        """
```

To follow our convention

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -324,3 +316,99 @@ def _set_name(self, name):
+        @param labels:
+        @param launcher:
+        @return:
+        '''
```
```suggestion
        """
```

To follow our convention

dimberman committed:
Update airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

kaxil commented on airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py:
```diff
@@ -324,3 +316,99 @@ def _set_name(self, name):
+        Montitors a pod to completion that was created by a previous KubernetesPodOperator
+
+        :param launcher:
+        :param pod:
```
Same here

dimberman committed:
Update airflow/providers/cncf/kubernetes/operators/kubernetes_pod.py

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>

astro-sql-decorator committed:
add docs

kaxil commented on airflow/kubernetes/pod_launcher.py:
```diff
@@ -116,9 +114,15 @@ def run_pod(
 
-    def _monitor_pod(self, pod: V1Pod, get_logs: bool) -> Tuple[State, Optional[str]]:
+        :param pod: pod spec that will be monitored
+        :@type V1Pod:
```
```suggestion
        :type pod : V1Pod
```

dimberman committed:
Update airflow/kubernetes/pod_launcher.py

Co-authored-by: Kaxil Naik <kaxilnaik@gmail.com>
